packages/block-editor/src/components/rich-text/use-enter.js

```diff
@@ -16,20 +21,47 @@ import { splitValue } from './split-value';
 
 export function useEnter( props ) {
 	const registry = useRegistry();
-	const { __unstableMarkAutomaticChange } = useDispatch( blockEditorStore );
+	const { insertBlock, insertDefaultBlock, __unstableMarkAutomaticChange } =
+		useDispatch( blockEditorStore );
+	const {
+		canInsertBlockType,
+		getBlockIndex,
+		getBlockName,
+		getBlockRootClientId,
+		getSelectedBlockClientId,
+	} = useSelect( blockEditorStore );
+	const { getDefaultBlockName } = useSelect( blocksStore );
 	const propsRef = useRef( props );
 	propsRef.current = props;
 	return useRefEffect( ( element ) => {
 		function onKeyDown( event ) {
-			if ( event.target.contentEditable !== 'true' ) {
+			if ( event.defaultPrevented ) {
 				return;
 			}
 
-			if ( event.defaultPrevented ) {
+			if ( event.keyCode !== ENTER ) {
 				return;
 			}
 
-			if ( event.keyCode !== ENTER ) {
+			if ( event.target.contentEditable !== 'true' ) {
+				const clientId = getSelectedBlockClientId();
+				const canInsertDefaultBlock = canInsertBlockType(
+					getDefaultBlockName(),
+					getBlockRootClientId( clientId )
+				);
+				if ( canInsertDefaultBlock ) {
+					insertDefaultBlock(
+						{},
+						getBlockRootClientId( clientId ),
+						getBlockIndex( clientId ) + 1
+					);
+				} else {
+					insertBlock(
+						createBlock( getBlockName( clientId ), {} ),
+						getBlockIndex( clientId ) + 1,
+						getBlockRootClientId( clientId )
+					);
```

Comment:
This doesn't feel right. Why isn't the block level Enter logic executing?

Reply:
I just tested this, and it actually does work without this extra logic in the `use-enter.js`. The only problem is it duplicates the button content:

<img width="514" alt="Screenshot 2024-02-26 at 1 13 58 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/96fa4cef-15f7-4b3c-8e81-1294ec4c7994">

If we're alright with forcing the `tabindex` to 0 when `shouldDisableEditing` is true, then we should be able to fix that duplication by modifying the [button block's `onSplit` prop](https://github.com/WordPress/gutenberg/blob/53c3d708c4d4c08856f6f43fe375a1a475960995/packages/block-library/src/button/edit.js#L296-L301)

@SantosGuillamot

Reply:
That happens when you remove [the whole conditional](https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-editor/src/components/rich-text/use-enter.js#L24-L26) that was introduced to fix the paragraph problem: [PR](https://github.com/WordPress/gutenberg/pull/59320).

From my testing, there are two "onKeyDown" functions running for the button, but none of them add the new button properly:

* [Rich text onKeyDown function](https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-editor/src/components/rich-text/use-enter.js#L23): It stops because of the mentioned conditional. Because `contentEditable` is false.
* [Button onKeyDown function](https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-library/src/button/edit.js#L68): It stops because the content is not empty at [this point](https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-library/src/button/edit.js#L73).

I assume it would be possible to move that logic to the button `onKeyDown` function somehow.

Reply:
I mean having adding something similar to this logic in [the `useEnter` of the button block](https://github.com/WordPress/gutenberg/blob/7eb0a5acb3f8ab7e8edb30e26934e25477fb1b67/packages/block-library/src/button/edit.js#L73):

```js
if ( metadata?.bindings?.text ) {
  insertBlock(
    createBlock( 'core/button', {} ),
    getBlockIndex( clientId ) + 1,
    getBlockRootClientId( clientId )
  );
  return;
}
```

Reply:
I meant the on enter logic defined by the all blocks: 

https://github.com/WordPress/gutenberg/blob/98552b2486b52e4dd3c90775c2a029b95e64253c/packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js#L46

It probably doesn't work because it tries to insert a default block. But inside a buttons block the only insertable block is a button blocks, so the default block should be button

Reply:
Okay, thanks for the info 🙂 

I can confirm that, removing the applied changes, it seems the issue is caused because, at [this point](https://github.com/WordPress/gutenberg/blob/53c3d708c4d4c08856f6f43fe375a1a475960995/packages/block-editor/src/store/actions.js#L1324), `getDefaultBlockName` return `core/paragraph`. And when it tries to insert the block it doesn't work because it expects a button (I guess).

If I hardcode it to "core/button", it seems to work.

Do you have any clue how I can change that for the button?

Reply:
I was considering using something like the following in the `insertDefaultBlock` action:

```js
const { name: directInsertBlockName } = rootClientId
			? select.getDirectInsertBlock( rootClientId )
			: null;

		if ( ! defaultBlockName && ! directInsertBlockName ) {
			return;
		}

		const block = createBlock(
			directInsertBlockName ? directInsertBlockName : defaultBlockName,
			attributes
		);
```

Would something like that make sense? Should it go to another place? Sorry for so many questions 😄 I'm totally new to this part of the codebase.

Reply:
Or maybe we could move the logic to [the on enter logic](https://github.com/WordPress/gutenberg/blob/53c3d708c4d4c08856f6f43fe375a1a475960995/packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js#L61):

```js
const rootClientId = getBlockRootClientId( clientId );
const index = getBlockIndex( clientId );
const directInsertBlock = rootClientId
  ? getDirectInsertBlock( rootClientId )
  : null;
if ( directInsertBlock ) {
  const block = createBlock(
    directInsertBlock?.name,
    {}
  );
  insertBlock( block, index + 1, rootClientId );
} else {
  insertDefaultBlock( {}, rootClientId, index + 1 );
}
```